### DISCUSSION TITLE
Any plans to develop `Materialized View` feature?

### DISCUSSION THREAD

**Discussion by xnge** [duckdb/duckdb#3638] [Ideas]
Hi, The materialized view is an important feature, Is there any plan to develop the feature of `materialized view`?
like this, https://www.postgresql.org/docs/current/rules-materializedviews.html

I know I can use `CREATE TABLE mymatview AS SELECT * FROM mytab;` to implement some functions.
But there is no ability to refresh data in this way, `REFRESH MATERIALIZED VIEW mymatview;`
Or is there any alternative way to do it?

**Comment by xnge** [comment]
hi @Mytherin ,

I'm planning to implement `Materialized View` in duckdb by myself and I'm going to do it this way, please help me review and give better suggestions.
1. `create materialized view` is very similar to `create table as xxx` in terms of program processing flow.
2. I can reuse the grammar here, with a few modifications. (create_as.y)[https://github.com/duckdb/duckdb/blob/master/third_party/libpg_query/grammar/statements/create_as.y]
3. add a new type of Materialized View in catelog.
4. execute the select statement of the materialized view to write the data to the materialized view storage when the view is refreshed  `REFRESH MATERIALIZED VIEW mymatview;`.

I am a little confused about the fourth point, if select is simpler, e.g.
```
CREATE MATERIALIZED VIEW mymatview AS SELECT * FROM mytab;
```
This way the Materialized View data only needs to be appended and written.

If it is more complex group by, for example
```
CREATE MATERIALIZED VIEW mymatview AS SELECT id, sum(price) FROM mytab GROUP BY id;
```
I still need to update the previous data, do you have any good suggestions?

**Reply by Mytherin** [reply]
Hey @xnge, thanks for the interest in implementing this feature!

Incremental materialized view maintenance of arbitrary queries is an incredibly complex topic. Postgres has [many restrictions on what types of SQL queries can and cannot be incrementally maintained](https://wiki.postgresql.org/wiki/Incremental_View_Maintenance). If you were to implement materialized views, I would start off by re-executing the entire query on `REFRESH`, and adding support for incremental view maintenance of more complex queries in steps.

In principle the battle plan seems alright to me. I would make sure to copy Postgres syntax and behavior as much as possible.

**Reply by xnge** [reply]
hi @Mytherin Thank you for your prompt response.
I am very happy to develop this feature that.
I plan to start with simple materialized views such as
```
CREATE MATERIALIZED VIEW mv1 AS SELECT * FROM tab1;
```
I hope I can get your help. Thanks!

**Reply by david542542** [reply]
@xnge what a great feature to implement, we'd use that a lot too! How's your progress coming along?

**Reply by Admolly** [reply]
I second that!  There's an interesting write-up [here](https://dax.tips/2019/11/11/creative-aggs-part-v-incremental-aggs/) on one method to "refresh" aggregate tables that could be worth considering.  Would be a great feature to have.

**Reply by code1704** [reply]
+1 for this feature

**Reply by Tishj** [reply]
Unless I'm misunderstanding, could you not use the `CREATE OR REPLACE` functionality to "refresh" the materialized view?

**Reply by marhar** [reply]
@Tisj, that would work in that the data state would be refreshed, but one of the nifty features of materialized views is that they can be continually updated as the source data changes, so that there's not a single big performance hit.

**Reply by WanenGao** [reply]
@xnge Hi, how's your project going?

**Reply by liuq2008202** [reply]
+1 for this feature

**Reply by jovezhong** [reply]
look forward to incremental materialized view or refreshable materialized views in DuckDB.

If you need something today to stream data from Kafka to S3/Iceberg, you may try the materialized views in https://github.com/timeplus-io/proton It's written in C++ and support similar features as Apache Flink. The materialized view is based on the query, not just the new data on the left side of the JOIN, so it supports global aggregation, tumble/hop/session window aggregation, ASOF join, etc.

**Comment by include** [comment]
+1 for this. Materialized view are a must to have, and as @marhar said above, "materialized views is that they can be continually updated as the source data changes" would be priceless :) (afaik Clickhouse has this).

**Comment by sirinath** [comment]
This  would be an pretty important feature to have.

**Comment by JingDas05** [comment]
I agree that incremental refresh for materialized views is a very useful feature. But once a materialized view is created, is transparent query rewriting needed?​​

For example, creating a materialized view like this:
```sql
CREATE MATERIALIZED VIEW mymatview AS 
SELECT id, date, sum(price) as sum_price 
FROM mytab 
GROUP BY id, date;
```

Then, a query such as:
```sql
SELECT id, sum(price) FROM mytab GROUP BY id;
```

Could be transparently rewritten as:
```sql
SELECT id, sum(sum_price) FROM mymatview GROUP BY id;
```
Modern OLAP engines generally support this capability.

**Comment by hpvd** [comment]
there is a fresh and very powerful approach to materialized views in turso (highly professional sqlite rewrite):
**Live Matrialized View**s,
where on update only the delta is calculate using DBSP Framework
see https://turso.tech/blog/introducing-real-time-data-with-materialized-views-in-turso

**Reply by Shonke** [reply]
This looks quite interesting

**Reply by hpvd** [reply]
jep, these guys do in general an amazing job with their rewrite... speed, quality (Deterministic Simulation Testing), Feature set..

**Comment by hpvd** [comment]
here is an idea how to make first build of materialized view with DBSP efficient
https://github.com/tursodatabase/turso/issues/3464

**Comment by hpvd** [comment]
just today there was a blog post about Streaming and DuckDB with a section about materialized view
https://duckdb.org/2025/10/13/duckdb-streaming-patterns.html#materialized-view-pattern-cooking-our-own-materialized-view-with-duckdb

and it includes a hint to the Roadmap with the topic Materialized View linking to this discussion https://duckdb.org/roadmap.html

**Comment by wuqunfei** [comment]
A perosnal prototype (DO NOT USE FOR PRODUCTION):  **A Declarative Pipeline, Materialized View for Duckdb with Data Quality Expectations.**
https://github.com/neuesql/hortus-pipeline

```sql
CREATE OR REFRESH MATERIALIZED VIEW <name>
  [CONSTRAINT <cname> EXPECT (<expr>) [ON VIOLATION {DROP ROW | FAIL UPDATE}]]
  [DEPENDS ON (<view1>, <view2>, ...)]
  [COMMENT '<text>']
  [SCHEDULE {EVERY <n> {SECOND|MINUTE|HOUR|DAY|WEEK} | CRON '<expr>' | TRIGGER ON UPDATE}]
AS <query>;
```

**Comment by ila** [comment]
fyi https://github.com/ila/openivm very much WIP but I'd be happy if anyone wants to play with it

**Reply by getravi** [reply]
I have been working on something similar at https://github.com/getravi/duckDBSP to power an application i am working on.
